grudge/dt_utils.py

```diff
@@ -0,0 +1,206 @@
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+
+import numpy as np
+
+from arraycontext import make_loopy_program
+
+from grudge.dof_desc import DD_VOLUME, DOFDesc
+from grudge.discretization import DiscretizationCollection
+
+import grudge.op as op
+
+from meshmode.dof_array import DOFArray
+
+from pytools import memoize_on_first_arg, memoize_in
+
+
+@memoize_on_first_arg
+def dt_non_geometric_factor(
+        dcoll: DiscretizationCollection, scaling=None, dd=None) -> float:
+    r"""Computes the non-geometric scale factor:
+
+    .. math::
+
+        C\operatorname{min}_i\left( \Delta r_i \right),
+
+    where :math:`\Delta r_i` denotes the distance between two distinct
+    nodes on the reference element, and :math:`C > 0` is a scaling
+    constant.
+
+    :arg scaling: a :class:`float` denoting the scaling factor. By default,
+        the constant is set to 2/3.
+    :arg dd: a :class:`~grudge.dof_desc.DOFDesc`, or a value convertible to one.
+        Defaults to the base volume discretization if not provided.
+    :returns: a :class:`float` denoting the minimum node distance on the
+        reference element.
+    """
+    if dd is None:
+        dd = DD_VOLUME
+
+    if scaling is None:
+        scaling = 2/3
+
+    discr = dcoll.discr_from_dd(dd)
+    min_delta_rs = []
+    for grp in discr.groups:
+        nodes = np.asarray(list(zip(*grp.unit_nodes)))
+        nnodes = grp.nunit_dofs
+
+        # NOTE: order 0 elements have 1 node located at the centroid of
+        # the reference element and is equidistant from each vertex
+        if grp.order == 0:
+            assert nnodes == 1
+            min_delta_rs.append(
+                np.linalg.norm(
+                    nodes[0] - grp.mesh_el_group.vertex_unit_coordinates()[0]
+                )
+            )
+        else:
+            min_delta_rs.append(
+                min(
+                    np.linalg.norm(nodes[i] - nodes[j])
+                    for i in range(nnodes) for j in range(nnodes) if i != j
+                )
+            )
+
+    # Return minimum over all element groups in the discretization
+    return scaling * min(min_delta_rs)
+
+
+@memoize_on_first_arg
+def dt_geometric_factor(dcoll: DiscretizationCollection, dd=None) -> float:
+    r"""Computes a geometric scaling factor, defined as the minimum radius of
+    an inscribed circle/sphere in a mesh cell.
+
+    Specifically, the inradius for each element is computed using the following
+    formula for simplicial cells (triangles/tetrahedra):
+
+    .. math::
+
+        r_{in} = \frac{d V}{\sum_{i=1}^{N_{faces}} F_i},
+
+    where :math:`d` is the topological dimension, :math:`V` is the cell volume,
+    and :math:`F_i` are the areas of each face of the cell.
```

Comment:
Should this be documented? As in, are we committing to this approach?

Reply:
> Should this be documented?

100% always yes.

> are we committing to this approach?

Well, I would say it's worth committing to this (it's certainly better than having _nothing_ ). There are objections to using `h_min_from_volume`, so this seemed like a reasonable alternative. What are the other alternatives here? Inradius seems to be a well-known and accepted mesh size parameter for polygonal meshes.

There are more elaborate constructions I'm sure we could entertain, but I'd rather not and just use something simple that works reasonably well.

Reply:
Fair enough. We're not really at liberty to change what this routine does without risking breakage, so we might as well document what it does on the inside.